### PR TITLE
Move discovery to separate functions.

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR removes the local vs global search option from `config.load_config` API in favor of a separate function to handle discovery. It also removes the `project_dir` key that was previously being automatically populated into the config when calling `load_config`.

## Motivation and Context

Populating the `project_dir` key is an unexpected side effect since it adds a key that doesn't actually exist in any config file. One option would be to change `load_config` to return a tuple containing `(root, config)` instead of just the config, but that mixes two distinct pieces of functionality and also makes config loading more expensive than it needs to be. By separating the two functions, calling code can find the config file and load it separately.

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [ ] New feature
- [x] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac/blob/master/changelog.txt) and added all related issue and pull request numbers for future reference (if applicable). See example below.

### CODE REVIEW

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:58]: We should create an absolute path once, before entering the `while True:`, then use `os.path.dirname` each time we want to go up. Performance is better if we (1) use `dirname` to go up instead of a join with `".."` and (2) call `abspath` fewer times.

```python
root = os.path.abspath(root)
while True:
    # ...
    up = os.path.dirname(root)
    # ...
```

Comment (bdice) [reply signac/common/config.py:58]: https://github.com/bdice/python_microbenchmarks/blob/main/microbenchmarks/os_path_dirname.py

Relative cost of each function call:
```
dirname: 0.0445243
join: 0.0719918
abs-dirname: 0.176682
abs-join: 0.234717
```

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:48]: Do we need/want to use [`os.path.expanduser`](https://docs.python.org/3/library/os.path.html#os.path.expanduser) or [`os.path.expandvars`](https://docs.python.org/3/library/os.path.html#os.path.expandvars) in any of our path handling? I think the answer is no (users can do that if needed), but I wanted to ask anyway.

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:69]: It looks like an extra indentation was added to the next few lines. The return section was also messed up?

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:102]:
```suggestion
    """Load configuration from a project directory.
```

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:107]: This sounds like it expects a config path, but I think it takes a project path. Also the use of the word "local" feels a bit out of place.

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:114]: Modifications to a `Config` object are _never_ reflected in a file, right? It's not like a synced collection iirc, but has methods to write to a file? Can't remember.

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:118]: We need to rename `cfg` to indicate it's a module-level constant, like `CONFIG_SCHEMA` or similar. It took me a moment to recognize what was going on here.

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:120]: Slightly redundant...?
```suggestion
    # Add in any global or user config files. For now this only finds user-specific
```

File: signac/contrib/project.py
Comment (bdice) [inline signac/contrib/project.py:1622]: Simplify this. I think this is equivalent? Maybe there's something even better in between the current state and this rewrite. My primary goal is to raise the `LookupError` only in one path.
```python
config_dir = _locate_config_dir(root) if search else root
if config_dir is None or not _contains_config_file(config_dir):
    raise LookupError(f"Unable to find project at path '{os.path.abspath(config_dir)}'.")
return cls(root=config_dir, **kwargs)
```

**Review by bdice — CHANGES_REQUESTED** [review]
First round of feedback. Thanks for working on this. :)

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:101]: Does this file need to have _any_ public methods?

**Comment by vyasr** [comment]
This is odd. A number of my changes got discarded somehow, leading to some of the weird half-done work that you commented on. Doesn't just look like uncommitted changes either. Not sure what happened, but in any case I'll fix and push. Most of your requests were changes I thought I made :(

File: signac/common/config.py
Comment (vyasr) [reply signac/common/config.py:58]: Good call.

File: signac/common/config.py
Comment (vyasr) [reply signac/common/config.py:107]: See if the new wording is clearer to you.

File: signac/common/config.py
Comment (vyasr) [reply signac/common/config.py:114]: The `configobj` API is pretty confusing in this respect. You are correct in the sense that it will not automatically reflect changes to a file. However, there `ConfigObj.write` behaves differently depending on whether you initialize the object with a filename or not. If you do, then `write` will create if necessary and write to the file. Otherwise, if you just do something like `c = ConfigObj(); c['a'] = 'b'; c.write()` it will just print the output. That means that a `Config` returned from `load_config` can't be written back to file, while a `Config` returned from `read_config_file` can be. I've been bitten by this many times in the past, so I wanted to add a note. I am open to a more robust solution in the long term that avoids this though, and I'm hopeful that whatever package we use for TOML support when we switch will allow us to avoid this problem. Alternatively, maybe in 2.0 we can remove all functions that return writeable configs and force any mutating code path to invoke the underlying TOML library directly. I'm not sure what the best approach is here, but for now I figured documentation was a start.

Comment (bdice) [reply signac/common/config.py:114]: Sounds good to me.

File: signac/common/config.py
Comment (vyasr) [reply signac/common/config.py:118]: Agreed. It is a public attribute, though, which isn't something that's really feasible to deprecate with a warning. I think we'll just have to rename this in 2.0. Chances are nobody is using it, but I know signac-flow used to (I believe that I removed that dependence recently when I removed monkey-patching), which is why I'm advocating treating it as a breaking change for 2.0 rather than something to put in this PR.

File: signac/contrib/project.py
Comment (vyasr) [reply signac/contrib/project.py:1622]: The problem is that the error message becomes useless in the case that `_locate_config_dir` returns `None`. I disliked this code as well but couldn't come up with an alternative that gets around that elegantly. You could do something like this:
```
old_root = root
root = _locate_config_dir(root) if search else root
if root is not None or not os.path.isfile(_get_project_config_fn(root)):
    raise LookupError(
        f"Unable to find project at path '{os.path.abspath(old_root)}'."
    )
```
which gets around my above objection, but it still has the downside of adding an extra pair of `os` calls in the `search` case because you'll run the `os.path.isfile(_get_project_config_fn(root))` even though the current value of `root` (the output of `_locate_config_dir`) is already sufficient information. I don't think that's worthwhile. The best we can do is just set a local `err` variable in both branches and raise the `LookupError` outside the `if-else`, which sort of addresses your concern but still isn't great.

File: signac/common/config.py
Comment (vyasr) [reply signac/common/config.py:101]: That is a fair question. Ultimately this boils down to a question of how we expect users to interact with config files in signac 2.0. The files will be hidden, so most casual users probably won't think to do something like
```
import toml  # or whatever package
config = toml.open("/path/to/project/.signac/config") # or whatever the API looks like
```
We could really expect them to treat it like git and edit directly with a text editor, but that's also not really entry-level. I only muck with the git config manually if I really have to. That leaves the command-line config interface. I'm fine with that being the primary entry point as it currently is, the question is whether we're OK with that being the _only_ entry point. Is that what we want?

File: signac/__main__.py
Comment (vyasr) [reply signac/__main__.py:413]: Good point, I obviously didn't really think this through. I'm not sure what to do with this. In the case where the workspace is a direct subdirectory of the project we could still support this (for the lifetime of 1.x) by doing
```
try:
    destination = Project(os.path.dirname(os.path.relpath(args.destination)))
except LookupError:
    ...
```
Is that worthwhile to avoid a breaking change here?

File: signac/common/config.py
Comment (vyasr) [reply signac/common/config.py:48]: I'm inclined to agree, let users do that. It's simpler and saves us from incurring the cost unconditionally. However, this point may be moot after we decide what to do on your last comment about whether any of these should be public APIs.

File: signac/common/config.py
Comment (bdice) [reply signac/common/config.py:101]: I would be fine with making it CLI / manual editing only. We have never recommended using these APIs, all our docs and error/warning messages point to using the CLI.

File: signac/contrib/project.py
Comment (bdice) [reply signac/contrib/project.py]: Ultimately this is pretty minor and I don’t want to get hung up on this. If you think one of these alternatives is better, go for it.

File: signac/__main__.py
Comment (bdice) [reply signac/__main__.py:413]: I personally think the “allow workspace” feature is a little unintuitive (or unexpected?) when everything else in signac is based on a project. There are no other functions I know of that deal with raw workspace directories. I’d be fine with removing it. I think @csadorf might have added this feature, so maybe helpful to get another opinion on whether we need it in 2.0.

Comment (bdice) [reply signac/__main__.py:413]: As for the lifetime of 1.x, this branch targets 2.0, right? I don’t see the connection. A breaking change is allowable.

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:23]: A better name might be `search_path`. The _returned_ value is the "root" of the project, not the argument here.
```suggestion
def _locate_config_dir(search_path):
```

File: signac/__main__.py
Comment (bdice) [reply signac/__main__.py:413]: It's been a couple days since I looked at this but I hadn't finished my review. I think the best action to take here is to remove `--allow-workspace` in a separate PR. We could do that first, or we can mark this as TODO and leave it partially broken with an issue to fix it later.

File: signac/common/config.py
Comment (bdice) [inline signac/common/config.py:60]: Are `configspec, *args, **kwargs` ever used? I don't see this function being called with anything except `filename` provided. Let's remove those arguments since they're not doing anything for us.

File: signac/common/config.py
Comment (bdice) [reply signac/common/config.py:86]: I recommend moving towards making all these functions private. Could be done in a later PR.

**Review by bdice — CHANGES_REQUESTED** [review]

File: signac/__main__.py
Comment (vyasr) [reply signac/__main__.py:413]: > As for the lifetime of 1.x, this branch targets 2.0, right? I don’t see the connection. A breaking change is allowable.
Yeah sorry I was context switching between too many PRs simultaneously when I wrote that. You're right, we can break this. I'll leave a TODO item for a future PR.

File: signac/common/config.py
Comment (vyasr) [reply signac/common/config.py:103]: Again, forgot this was a PR where breaks were fine. That said, I think this is another future TODO IMO.

File: signac/common/config.py
Comment (vyasr) [reply signac/common/config.py:86]: Sounds good, I'll keep that on the docket for future work.

File: signac/__main__.py
Comment (bdice) [inline signac/__main__.py:366]: Unrelated to this PR, but I think this is supposed to say:
```suggestion
    # Validate provided argument combinations
```

**Review by bdice — APPROVED** [review]
Nice. Let's merge this, so that we can unblock other work.

**Pull request merged**
